crates/extensions/c8y_firmware_manager/src/actor.rs

```diff
@@ -0,0 +1,756 @@
+fan_in_message_type!(FirmwareOutput[MqttMessage, OperationSetTimeout, IdDownloadRequest] : Debug);
+
+pub struct FirmwareManagerActor {
+    config: FirmwareManagerConfig,
+    active_child_ops: HashMap<OperationKey, ActiveOperationState>,
+    reqs_pending_download: HashMap<String, SmartRestFirmwareRequest>,
+    message_box: FirmwareManagerMessageBox,
+}
+
+impl FirmwareManagerActor {
+    pub fn new(config: FirmwareManagerConfig, message_box: FirmwareManagerMessageBox) -> Self {
+        Self {
+            config,
+            active_child_ops: HashMap::new(),
+            reqs_pending_download: HashMap::new(),
+            message_box,
+        }
+    }
+
+    pub async fn process_mqtt_message(
+        &mut self,
+        message: MqttMessage,
+    ) -> Result<(), FirmwareManagementError> {
+        if self.config.c8y_request_topics.accept(&message) {
+            self.handle_firmware_update_smartrest_request(message)
+                .await?;
+        } else if self.config.firmware_update_response_topics.accept(&message) {
+            self.handle_child_device_firmware_operation_response(message.clone())
+                .await?;
+        } else {
+            error!(
+                "Received unexpected message on topic: {}",
+                message.topic.name
+            );
+        }
+        Ok(())
+    }
+
+    pub async fn handle_firmware_update_smartrest_request(
+        &mut self,
+        message: MqttMessage,
+    ) -> Result<(), FirmwareManagementError> {
+        for smartrest_message in collect_smartrest_messages(message.payload_str()?) {
+            let result = match get_smartrest_template_id(&smartrest_message).as_str() {
+                "515" => match SmartRestFirmwareRequest::from_smartrest(&smartrest_message) {
+                    Ok(firmware_request) => {
+                        self.handle_firmware_download_request(firmware_request)
+                            .await
+                    }
+                    Err(_) => {
+                        error!("Incorrect c8y_Firmware SmartREST payload: {smartrest_message}");
+                        Ok(())
+                    }
+                },
+                _ => {
+                    // Ignore operation messages not meant for this plugin
+                    Ok(())
+                }
+            };
+
+            if let Err(err) = result {
+                error!("Handling of operation: '{smartrest_message}' failed with {err}");
+            }
+        }
+        Ok(())
+    }
+
+    async fn handle_firmware_download_request(
+        &mut self,
+        smartrest_request: SmartRestFirmwareRequest,
+    ) -> Result<(), FirmwareManagementError> {
+        info!(
+            "Handling c8y_Firmware operation: device={}, name={}, version={}, url={}",
+            smartrest_request.device,
+            smartrest_request.name,
+            smartrest_request.version,
+            smartrest_request.url,
```

Comment:
This formatting could be moved to some `impl Display for FirmwareManagementError`.